plugins/catalog-backend/src/next/types.ts

```diff
@@ -45,29 +44,29 @@ export interface LocationService {
   deleteLocation(id: string): Promise<void>;
 }
 
-export type EntityMessage =
-  | { all: Entity[] }
-  | { added: Entity[]; removed: EntityName[] };
-
 export interface LocationStore {
-  // extends EntityProvider
   createLocation(spec: LocationSpec): Promise<Location>;
   listLocations(): Promise<Location[]>;
   getLocation(id: string): Promise<Location>;
   deleteLocation(id: string): Promise<void>;
-
-  location$(): Observable<
-    { all: Location[] } | { added: Location[]; removed: Location[] }
-  >;
 }
 
 export interface CatalogProcessingEngine {
   start(): Promise<void>;
   stop(): Promise<void>;
 }
 
+export type EntityProviderMutation =
+  | { type: 'full'; entities: Entity[] }
+  | { type: 'delta'; added: Entity[]; removed: Entity[] };
+
+export interface EntityProviderConnection {
+  applyMutation(mutation: EntityProviderMutation): Promise<void>;
+}
+
 export interface EntityProvider {
-  entityChange$(): Observable<EntityMessage>;
+  getProviderName(): string;
```

Comment:
Was thinking that we should provide this together with the provider rather than baked into it, but realized that this will allows us to encode whether to allow multiple instances of the same provider or not. Thinking this:

```ts
builder.addProviders(
  new DefaultLocationStore(...),
  new GitHubOrgScanner({ org: 'org-1' }), // getProviderName() === 'GitHubOrgScanner_org-1'
  new GitHubOrgScanner({ org: 'org-2' }), // getProviderName() === 'GitHubOrgScanner_org-2'
)
```

Kinda think it should be called `getProviderId` tho x)

Reply:
A bit risky with implicit / computed IDs. If you add two of the same processor, both for the same target but with different filter parameters and setting different namespace on the output entities or whatever... Then you have to start adding the filters to the id as well etc

Maybe force the user to come up with IDs is safest